hwmon: (pmbus/max34440): Add support adpm12160

actorreno opened the pull request:
ASPM12160 is a quarter brick DC/DC Power Module. It is a high power non-isolated converter capable of delivering a fully regulated 12V, with continuous power level of 1600W with peak power at 2400W for a limited time. Uses PMBus Configuration.

## PR Description

- This adds ADPM12160 driver support on existing MAX34440 Driver.
- Datasheets [ADPM12160](websitehere).
- Tested on RPI4 using the following eval boards: [ADPM12160-EVB](websitehere).

## PR Type
- [ ] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

kister-jimenez commented on Documentation/hwmon/max34440.rst:
```diff
@@ -128,13 +137,17 @@ in[1-6]_highest		Historical maximum voltage.
 
 ======================= ========================================================
-curr[1-6]_label		"iout[1-6]".
+curr[1-6]_label		"iout[1-6]". For ADPM12160 curr[1] is "iin1" and curr[2-6]
```
Move the ADPM12160 comment in the note.

actorreno replied:
will move

kister-jimenez commented on drivers/hwmon/pmbus/max34440.c:
```diff
@@ -59,12 +63,20 @@ static int max34440_read_word_data(struct i2c_client *client, int page,
 
 	switch (reg) {
 	case PMBUS_IOUT_OC_FAULT_LIMIT:
-		ret = pmbus_read_word_data(client, page, phase,
-					   MAX34440_IOUT_OC_FAULT_LIMIT);
+		if (data->id != adpm12160)
+			ret = pmbus_read_word_data(client, page, phase,
+						   MAX34440_IOUT_OC_FAULT_LIMIT);
+		else
+			ret = pmbus_read_word_data(client, page, phase,
+						   ADPM12160_IOUT_OC_FAULT_LIMIT);
 		break;
 	case PMBUS_IOUT_OC_WARN_LIMIT:
-		ret = pmbus_read_word_data(client, page, phase,
-					   MAX34440_IOUT_OC_WARN_LIMIT);
+		if (data->id != adpm12160)
+			ret = pmbus_read_word_data(client, page, phase,
+						   MAX34440_IOUT_OC_WARN_LIMIT);
+		else
+			ret = pmbus_read_word_data(client, page, phase,
+						   ADPM12160_IOUT_OC_WARN_LIMIT);
```
Isn't it MAX34440 the only one using the incorrect command address? If yes, maybe we can swap the conditionals to check if it is MAX34440, use the MAX34440 commands, otherwise, use the correct address.

kister-jimenez replied:
@nunojsa The commands address in MAX34440 is incorrect at the die level and I think the other devices are using correct address. Should we have this as a separate commit to fix this?

actorreno replied:
the apps engr just replied:

"MAX34446 has the same issue, and it was not fixed by firmware updates.
For MAX34451, MAX34451ETNA4+ and earlier had the same issue. The issue was fixed in MAX34451ETNA6+."

so 34446 is the same as 34440 but I am not sure how to go about the 344451

nunojsa replied:
Ok, so the driver already has:

```
/*
 * The whole max344* family have IOUT_OC_WARN_LIMIT and IOUT_OC_FAULT_LIMIT
 * swapped from the standard pmbus spec addresses.
 */
```

So, the doubt is about MAX34451 where some variants have the issues and others don't? IIUC, the other sane way to fix it is to add a new compatible for MAX34451ETNA6 like max34451-na6 (something in these lines). And replying to @kister-jimenez, yes, if we do fix this for MAX34451, it needs to be a separate commit before adding support for the new device.

kister-jimenez commented on drivers/hwmon/pmbus/max34440.c:
```diff
@@ -133,12 +146,20 @@ static int max34440_write_word_data(struct i2c_client *client, int page,
 	case PMBUS_IOUT_OC_FAULT_LIMIT:
-		ret = pmbus_write_word_data(client, page, MAX34440_IOUT_OC_FAULT_LIMIT,
-					    word);
+		if (data->id != adpm12160)
+			ret = pmbus_write_word_data(client, page, MAX34440_IOUT_OC_FAULT_LIMIT,
+						    word);
+		else
+			ret = pmbus_write_word_data(client, page, ADPM12160_IOUT_OC_FAULT_LIMIT,
+						    word);
 		break;
 	case PMBUS_IOUT_OC_WARN_LIMIT:
-		ret = pmbus_write_word_data(client, page, MAX34440_IOUT_OC_WARN_LIMIT,
-					    word);
+		if (data->id != adpm12160)
+			ret = pmbus_write_word_data(client, page, MAX34440_IOUT_OC_WARN_LIMIT,
+						    word);
+		else
+			ret = pmbus_write_word_data(client, page, ADPM12160_IOUT_OC_WARN_LIMIT,
+						    word);
```
ditto

kister-jimenez reviewed (commented):
First pass review. I think you also need to update the Kconfig help section to include this part..

actorreno commented:
v2

added a new commit where the versions of max34451 (max34451etna6+ and later) is supported first
before the commit where the support for adpm12160 is added

nunojsa reviewed (commented):
LGTM... please send it upstream

actorreno commented:
noticed a typo in rst file, fixed it

will upstream, thanks!

actorreno commented:
v3: (upstream applied)

Commit 1: MAX34451 Fix
- uses MFR_REVISION to know if max34451 is version max34451etna6+ and above
- added variable in private struct to denote what pmbus address will be in use (wrong one or the correct one)
- added support for a few more pmbus commands supported by max34451etna6+
- removed changes in .rst as deemed unecessary
Commit 2: add support adpm12160
- just edited to match changes in commit 1.

nunojsa reviewed (approved):
Is this as-is upstream? Next time, directly cherry pick the upstream commit so I see the maintainer tags and then there's no doubt

actorreno commented:
> Is this as-is upstream? Next time, directly cherry pick the upstream commit so I see the maintainer tags and then there's no doubt

Yes this is as-is upstream
Will note the cherry pick next time around, thank you!
